pkg/gateway/watcher.go

```diff
@@ -0,0 +1,64 @@
+package gateway
+
+import (
+	"context"
+	"github.com/samber/lo"
+	"google.golang.org/protobuf/proto"
+)
+
+type ConditionWatcher interface {
+	WatchEvents()
+}
+
+type internalWatcher struct {
+	closures []func()
+}
+
+func NewConditionWatcher(cl ...func()) ConditionWatcher {
+	return &internalWatcher{
+		closures: cl,
+	}
+}
+
+func (w *internalWatcher) WatchEvents() {
+	for _, cl := range w.closures {
+		f := cl
```

Comment:
Nit: why are we creating a new variable here?

Reply:
If we don't create `f` we risk running the wrong function in the following closure since the goroutine would capture the current value of `cl` when the goroutine starts rather than at the time the goroutine was declared. So we need to copy the func to ensure that we are calling the right function in the goroutine.

https://go.dev/play/p/Q_OTVaYclXB

Reply:
The form `cl := cl` is more common in our codebase, that's what I would recommend

Reply:
@kralicky renamed :+1: